roda-core/roda-core/src/main/java/org/roda/core/storage/DefaultTransactionalStorageService.java

```diff
@@ -769,7 +789,69 @@ private void copyToStagingStorageService(StorageService fromStorage, StoragePath
   @Override
   public void rollback() throws RODATransactionException {
     LOGGER.warn("[transactionId:{}] Rolling back transaction", transaction.getId());
-    // DO NOTHING
+    List<TransactionStoragePathConsolidatedOperation> successfulOperations = transactionLogService
+      .getSuccessfulConsolidatedStoragePathOperations(transaction);
+    try {
+      for (TransactionStoragePathConsolidatedOperation operation : successfulOperations) {
+        transactionLogService.updateConsolidatedStoragePathOperationState(operation, OperationState.ROLLING_BACK);
+        if (operation.getOperationType() == OperationType.DELETE) {
+          // rollbackDeleteOperation(operation);
+        } else if (operation.getOperationType() == OperationType.UPDATE) {
+          rollbackUpdateOperation(operation);
+        } else if (operation.getOperationType() == OperationType.CREATE) {
+          rollbackCreateOperation(operation);
+        } else if (operation.getOperationType() == OperationType.READ) {
+          // do nothing for read operations
+        }
+        transactionLogService.updateConsolidatedStoragePathOperationState(operation, OperationState.ROLLED_BACK);
+      }
+    } catch (RODATransactionException e) {
+      for (TransactionStoragePathConsolidatedOperation operation : successfulOperations) {
+        transactionLogService.updateConsolidatedStoragePathOperationState(operation, OperationState.ROLL_BACK_FAILURE);
+      }
+      throw new RODATransactionException(
+        "[transactionId:" + transaction.getId() + "] Failed to rollback this transaction", e);
+    }
+  }
+
+  public void rollbackCreateOperation(TransactionStoragePathConsolidatedOperation operation)
+    throws RODATransactionException {
+    try {
+      StoragePath storagePath = FSUtils.getStoragePathFromString(operation.getStoragePath());
```

Comment:
We shouldn't use FSUtils here, as it depends on the storage implementation and should only be used for file system-based storages.